src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -0,0 +1,263 @@
+        }
+
+        @Override
+        public void create(ActionListener<CreateIndexResponse> actionListener) throws IOException {
+            if (isCreated()) {
+                return;
+            }
+
+            CreateIndexRequest request = new CreateIndexRequest(MODEL_INDEX_NAME)
+                    .mapping("_doc", getMapping(), XContentType.JSON)
+                    .settings(Settings.builder()
+                            .put("index.hidden", true)
+                            .put("index.number_of_shards", this.numberOfShards)
+                            .put("index.number_of_replicas", this.numberOfReplicas)
+                    );
+            client.admin().indices().create(request, actionListener);
+        }
+
+        @Override
+        public boolean isCreated() {
+            return clusterService.state().getRoutingTable().hasIndex(MODEL_INDEX_NAME);
+        }
+
+        @Override
+        public void put(String modelId, KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+            String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+            Map<String, Object> parameters = ImmutableMap.of(
+                    KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                    KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+            );
+
+            IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+            indexRequestBuilder.setId(modelId);
+            indexRequestBuilder.setSource(parameters);
+
+            put(indexRequestBuilder, listener);
+        }
+
+        @Override
+        public void put(KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+            String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+            Map<String, Object> parameters = ImmutableMap.of(
+                    KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                    KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+            );
+
+            IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+            indexRequestBuilder.setSource(parameters);
+
+            put(indexRequestBuilder, listener);
+        }
+
+        private void put(IndexRequestBuilder indexRequestBuilder, ActionListener<IndexResponse> listener) {
+            if (!isCreated()) {
+                throw new IllegalStateException("Cannot put model in index before index has been initialized");
+            }
+
+            // Fail if the id already exists. Models are not updateable
+            indexRequestBuilder.setOpType(DocWriteRequest.OpType.CREATE);
+            indexRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+            indexRequestBuilder.execute(listener);
+        }
+
+        @Override
+        public byte[] get(String modelId) throws ExecutionException, InterruptedException {
+            if (!isCreated()) {
```

Comment:
Having ```isCreated``` check for each get request could be adding latencies. Can we instead rely on Elasticsearch to throw exception if index not present and we avoid explicitly checking it? We could probably wrap in try catch and look for Index not found exception or similar one?

Reply:
`isCreated` reads from Cluster metadata to determine if index is created, so it is not very expensive. I was thinking it would be better to fail first before making a GET call. What do you think?

Reply:
IMO majority of the cases will have Model index available and this check might be unnecessary. So I still align towards depending on Elasticsearch to throw exception

Reply:
Makes sense. Updated `isCreated` for GET request.